### PR TITLE
remap-redux part 2

### PR DESCRIPTION
**Pull request by fr500**
Goals: 

- [x] allow multiple gamepads to support keymapper
- [x] add keymapper to vita
- [x] add sublabels with the current input device we're remapping in the keyboard mapper
- [x] invert the columns in the gamepad remapping interface
- [x] add sublabels with the current input device we're remapping in the gamepad mapper

Extra:
- [x] allow analog remapping
- [ ] add image previews

--------------------------

I think the goals are achieved, the images stuff will have to wait, partly because I'm tired, and partly because I don't want this code to get intertwined with XMB code.

There are still a few bugs I hope can be fixed:

- if the device in settings is RETRO_DEVICE_KEYBOARD and now you load a core that doesn't support keyboard you get the keymapper (this needs to be sorted as a whole in RetroArch, it has been an issue forever)
- Mapping several inputs to the same action with analogs involved cause unexpected results
- Mapping the enter key on the keymapper causes it to get stuck in a loop until you press another button. Doesn't seem to to happen with other keys
- Mapping some keys on the keymapper prevents those keys from working on the keyboard

What currently works:
- [x] Mapping keyboard keys from more than one gamepad (works with dosbox)
- [x] Mapping more than one button to the same action
- [x] Unmapping buttons
- [x] Unmapping analogs
- [x] Mapping a button to trigger an analog response (tested with mupen, can run on SM64 with the d-pad now, triggers a full analog tilt)
- [x] Mapping an analog to another analog (having more than one analog mapped to the same output causes issues)
- [x] Mapping an analog to produce a button response

### CODE REVIEW

File: Makefile.vita [hunk 1]
```diff
@@ -33,6 +33,7 @@ else
 	HAVE_ZLIB		:= 1
 	HAVE_7ZIP		:= 1
 	HAVE_VITA2D		:= 1
+   HAVE_KEYMAPPER := 1
```
Comment (RobLoach) [inline Makefile.vita]: TAB instead of four spaces.

File: input/input_mapper.c [hunk 2]
```diff
@@ -82,9 +83,10 @@ void input_mapper_free(input_mapper_t *handle)
 
 void input_mapper_poll(input_mapper_t *handle)
 {
-   int i;
+   int i, j;
    settings_t *settings = config_get_ptr();
    unsigned device      = settings->uints.input_libretro_device[handle->port];
+bool key_event[RARCH_CUSTOM_BIND_LIST_END];
```
Comment (RobLoach) [inline input/input_mapper.c]: Whitespace

File: input/input_mapper.c [hunk 3]
```diff
@@ -100,23 +102,33 @@ void input_mapper_poll(input_mapper_t *handle)
 #endif
 
    memset(handle->keys, 0, sizeof(handle->keys));
-
-   for (i = 0; i < RARCH_CUSTOM_BIND_LIST_END; i++)
+   i = 0;
+   for (i = 0; i < 8; i++)
```
Comment (RobLoach) [inline input/input_mapper.c]: What does 8 represent here?

**Comment by fr500** [comment]
8 as random number of pads for testing, as it is, it allows the keyboard mapper to work with 8 users :P
I have to query the actual number of users

Also I need sublabels to identify the user device being mapped to keyboard because of this:

image

image

I was thinking left column should be

DualShock 3 Controller:  -------------->label
Triangle (btn) ------------------------->sublabel

Right column would be intact for keymapper but I think for the gamepad mapper I want to invert the columns and then right column would be

SNES Gamepad: B button -------------> single line

I was thinking using the left / right artwork stuff @Tatsuya79 is working on but there isn't all that much space to begin with...

image

The idea would be to have something like this

image

**Comment by fr500** [comment]
Woops just realized my pad was a DS3 and I used a xbox pad there for reference
More accurate representation of the idea:

image

**Comment by fr500** [comment]
The way this would work is:

Alongside with the current theme files there would be controller skins folder with the images for the pads. These images would be like this:

```
---snes\
-------base.png
-------a.png
-------b.png
-------x.png
-------etc
---nes\
-------base.png
-------a.png
-------b.png
-------etc
---ds4\
-------base.png
-------triangle.png
-------circle.png
-------etc
```

Then we would draw base and... B on top of base (so the skins take less space)
Of course building up the names to draw and all of that is gonna be a nightmare, but I'm pretty sure it can be done.

**Comment by fr500** [comment]
There would need to be a generic gamepad depiction too so we don't end up with empty space in that case.

**Comment by markwkidd** [comment]
@herbfargus I know this doesn't really affect RetroPie, but aren't you the keeper of some lore about RetroArch gamepad imagery?

**Comment by HerbFargus** [comment]
I worked on a project of sorts with @taalas we kept the archives here:

https://github.com/taalas/RetroPie-Artwork/tree/master/Source_Graphics

I used them to generate most of the diagrams in the retropie docs. He drew all of them, I just added the button correlations from retropad.

**Comment by markwkidd** [comment]
That was it, thanks!

**Comment by taalas** [comment]
@markwkidd @HerbFargus Any way I can help? I would gladly try to provide more artwork if needed...

File: menu/cbs/menu_cbs_get_value.c [hunk 4]
```diff
@@ -631,15 +631,22 @@ static void menu_action_setting_disp_set_label_input_desc_kbd(
...
+   unsigned offset = 0;
+
    settings_t *settings = config_get_ptr();
 
    if (!settings)
       return;
 
+   offset = type / ((MENU_SETTINGS_INPUT_DESC_KBD_END - (MENU_SETTINGS_INPUT_DESC_KBD_END - MENU_SETTINGS_INPUT_DESC_KBD_BEGIN))) - 1;
+
+   id = (type / (offset + 1)) - MENU_SETTINGS_INPUT_DESC_KBD_BEGIN;
```
Comment (fr500) [inline menu/cbs/menu_cbs_get_value.c]: btw I have no idea of what the math here means, it just worked so I settled with it

**Comment by Jaffacakelover** [comment]
That pic where you're showing 360/PS3 controllers... Could there be a line in a controller autoconfig file to define what image it displays (or what console its based on)? It'd be extra nice, for example, if the Mayflash N64 configs showed an N64 controller instead of a 360.

**Comment by fr500** [comment]
@Jaffacakelover well you should have read the description of what I intend to do.

**Comment by fr500** [comment]
Some progress with the keyboard part today, finished the keymapper basically (might be some bugs), it only shows keymapper entries for devices with a kbd enabled (or a subclass) and adds a handy sublabel to help identify what device we're working with (I still dislike the fact that we refer players as Users but whatever)

image

image

**Comment by fr500** [comment]
Now I have to invert the columns for the gamepad mapper

**Comment by Jaffacakelover** [comment]
@fr500 Believe it or not, I did read through before commenting (I probably didn't catch the jargon). If the feature I'm thinking of is already going in (picking up a line like "padstyle=nes" in autoconfig files), cool, thanks!

**Comment by fr500** [comment]
Refined it a bit further:

image

image

I just realized the new gui would allow setting buttons to more than one action (at least visually) but the logic still has to be implemented

The gamepad mapper is only tweaked visualy atm, need to hookup left / right callbacks and then work on the logic but that's I can do right now.

**Comment by fr500** [comment]
I can't do this, the input_state / input_poll code is too difficult for me to fix the gamepad mapper.

**Comment by twinaphex** [comment]
State specifically and clearly what the issues are and what you need help with that you cannot figure out, and I will take a look.

**Comment by fr500** [comment]
Ok, I'm gonna be very thorough:

This is how the input remapping gui looks like now (in master)
image

This GUI has several limitations as outlined in several issues including:
https://github.com/libretro/RetroArch/issues/3671

The most important problems:

- It doesn't allow to unbind buttons
- It doesn't allow to have extra gamepad buttons used for N:1 mapping (even though you have extra buttons you can't use them because the GUI makes this impossible, the GUI is limited by the descriptors set by the core, if the core sets 4 descriptors, the GUI will present 4 remappable buttons
- It doesn't allow to remap analogs
- It doesn't show you a reference of what you're remapping

Now, the new GUI looks like this:

image

As off now, the gui is fully functional, except for some error checking and for the missing "---" label on empty maps (which is already implemented for keyboard:

image

Now for the problems:

The old gamepad mapper approach worked by changing the pressed button id on input_state.c as follows:

```C
      if (settings->bools.input_remap_binds_enable)
      {
         switch (device)
         {
            case RETRO_DEVICE_JOYPAD:
               if (id < RARCH_FIRST_CUSTOM_BIND)
                  id = settings->uints.input_remap_ids[port][id];
               break;
            case RETRO_DEVICE_ANALOG:
               if (idx < 2 && id < 2)
               {
                  unsigned new_id = RARCH_FIRST_CUSTOM_BIND + (idx * 2 + id);

                  new_id = settings->uints.input_remap_ids[port][new_id];
                  idx   = (new_id & 2) >> 1;
                  id    = new_id & 1;
               }
               break;
         }
}
```

It doesn't check if the button is remapped or anything, basically overwrites everything with the new button press.
This works, and works very well except for the issues mentioned above.

So for the new GUI I had to not only change how the values are printed but what the data structures represent.

While both old and new remaps look identical (example below)

```
input_player1_b = "8"
input_player1_y = "0"
input_player1_a = "9"
input_player1_x = "1"
```

What these columns represent is different, in the old case the key represents the core button, and the value represents the gamepad button.

In the new one the key represents the gamepad button (mapped to retropad) and the value represents the core button.

This had to be done, because there is no way to change the values otherwise, the previous menu representation referenced this:

```C
description = system->input_desc_btn[p][desc_offset];
```

It's constant, can't be increased / decreased and it basically represents the label which has no significance.

So now, the actual issue with the current code:

To make this work for keyboard I added a new file, input_mapper.c, which is supposed to implement all the input mangling we might do:

- Turbo
- Remapping
- Combos
- N:1 mapping

The problem right now is that the current implementation won't work, you can't just change the id, what you need to do is build a new input_state based on the old input_state (press the new "virtual" buttons" and pass them to the core)

Because I have to do this from input_state itself I get one of these behaviors no matter what I try:

a. I get duplicated inputs on remapped buttons (because the old button state still exists for the previous button)

b. I get no imputs on the remapped buttons

To test this you can use this very simple remap with FCEUMM:

```
input_player1_b = "8"
input_player1_y = "0"
input_player1_a = "10"
input_player1_x = "10"
input_libretro_device_p1 = "1"
input_player1_analog_dpad_mode = "0"
input_libretro_device_p2 = "1"
input_player2_analog_dpad_mode = "0"
input_libretro_device_p3 = "1"
input_player3_analog_dpad_mode = "0"
input_libretro_device_p4 = "1"
input_player4_analog_dpad_mode = "0"
```

What it does is maps core BA to retropad YB  (common remap) and maps turbo buttons to some non-user inputs.

**Comment by fr500** [comment]
a. happens because the input state from the original input still exists
b. happens when I clear that input state, the problem is that the input_mapper_poll function uses that state to find out if a button has to be pressed.

I don't really know how to get the buttons pressed in input_poll directly, if I knew then I could use that information instead of calling input_state here:

https://github.com/libretro/RetroArch/pull/6485/files#diff-3aca7abba9e983b2b9da64aae080ecbaR148

Maybe if I can get the button state directly by here:
https://github.com/libretro/RetroArch/pull/6485/files#diff-7797ae0e7813aded66f109c460a562daL616

I could pass it to input_mapper_poll and avoid relying on input_state

**Comment by fr500** [comment]
image

After banging my head against the desk all day, the new mapper works, also N:1 mapping

**Comment by fr500** [comment]
So back to the buttons, maybe instead of just showing the active one, we can color code them

image

image

To allow understanding the mapping at a glance

**Comment by fr500** [comment]
Another option would be something like this:

image
Definitely uglier tho

**Comment by fr500** [comment]
@bibalox would you be interested in commenting on this issue?

The idea I have atm is to have a simple skinning system for controllers consisting of a base:

image

And a colored overlay

image

Or a label

image

Not sure which would work better.
The idea would be to draw on top of the base to get the desired effect

**Comment by fr500** [comment]
Steam:

image

image

**Comment by Tatsuya79** [comment]
You could use the space on the side to add the label:
pad

(random color here)

**Comment by fr500** [comment]
@twinaphex @bparker06 

I need help with a part of this implementation:

This is the code that writes the new input:
```C
         for (j = 0; j < RARCH_FIRST_CUSTOM_BIND; j++)
         {
            current_button_value = BIT256_GET(current_input, j);
            remap_button = settings->uints.input_remap_ids[i][j];
            if (current_button_value == 1 && j != remap_button &&
                remap_button != RARCH_UNMAPPED)
               BIT256_SET(handle->buttons, remap_button);
         }
```

Doesn't seem to work for any players other than P1. 
The code in input_state() resets the original input successfully but this code is not writing the new fake input for the pressed button correctly.

**Comment by fr500** [comment]
Actually I see part of the problem:

```C
      input_keys_pressed(settings, &current_input);
      BIT256_CLEAR_ALL(handle->buttons);
      for (i = 0; i < MAX_USERS; i++)
      {
         /* this loop iterates on all users and all buttons, and checks if a pressed button
            is assigned to any other button than the default one, then it sets the bit on the
            mapper input bitmap, later on the original input is cleared in input_state */

         for (j = 0; j < RARCH_FIRST_CUSTOM_BIND; j++)
         {
            current_button_value = BIT256_GET(current_input, j);
            remap_button = settings->uints.input_remap_ids[i][j];
            if (current_button_value == 1 && j != remap_button &&
                remap_button != RARCH_UNMAPPED)
            {
               BIT256_SET(handle->buttons, remap_button);
               RARCH_LOG("remap i:%d j:%d n:%d\n", i, j, remap_button);
            }
         }
```

`current_button_value = BIT256_GET(current_input, j);`

Is only taking **j** in account, not **i**. 

I tried **i * j** initially but that breaks even player 1. how should I offset **j** to take other players into account

Does input_keys_pressed(settings, &current_input);` even get values for users other than 1?

**Comment by bparker06** [comment]
@fr500 It looks like you're hard-coding the user to 0 here:

https://github.com/fr500/RetroArch/blob/d01c6359a1269bf46d2f191f1355cef31d67b85d/input/input_mapper.c#L120

**Comment by fr500** [comment]
Nope

```int16_t input_state(unsigned port, unsigned device,
      unsigned idx, unsigned id);```

idx is the analog stick, left or right, also that's the keyboard mapper

**Comment by bparker06** [comment]
Hmm I was going off of this:
```
 * input_state:
 * @port                 : user number.
 * @device               : device identifier of user.
 * @idx                  : index value of user.
 * @id                   : identifier of key pressed by user.
```
Not really sure even what the difference of all those are.

**Comment by fr500** [comment]
Yeah saw that too, but that's not the problem
actually input_keys_pressed does only work for P1... sucks

Guess I need a function much like that one to get the state of a specific player (without all the hotkey mumbo jumbo probably)

Still even if I can tell if a P2 or P3 or whatever button is pressed, then I need to set the bits with the correct offsets for this to work right?

**Comment by fr500** [comment]
@twinaphex 
The old mapper worked like this: change the pressed button before passing it to the core, it didn't care  about anything else, and that's why it couldn't do anything else other than change ids.

It couldn't do N:1 mapping, it couldn't even unmap a single button, and it couldn't possible ever remap analogs to buttons and viceversa.

What this mapper does is reads a single button input state, and rewrites it with any other input that can be implemented, it could do turbo per button, it could do 1:M mapping (combos), it already allows unmapping buttons, and N:1 mapping (multiple physical buttons can perform the same retropad action), it could do analog to digital (it's not implemented and explained why and how it should be done in the code) and digital to analog.

What doesn't work right now (reasons explained above): 

- P2+ remapping is broken, I asked for help about that above, should be simple-ish to address once I understand the stuff I just asked
- Analog to digital and Digital to analog, it needs a simple way to query the whole analog state for a player, pretty much the same issue above.

The code that does the remapping now is quite simple:

```
   if (device == RETRO_DEVICE_JOYPAD)
   {
      BIT256_CLEAR_ALL(handle->buttons);
      for (i = 0; i < 2; i++)
      {
         input_keys_pressed(settings, i, &current_input);
         /* this loop iterates on all users and all buttons, and checks if a pressed button
            is assigned to any other button than the default one, then it sets the bit on the
            mapper input bitmap, later on the original input is cleared in input_state */

         for (j = 0; j < RARCH_FIRST_CUSTOM_BIND; j++)
         {
            current_button_value = BIT256_GET(current_input, j);
            remap_button = settings->uints.input_remap_ids[i][j];
            if (current_button_value == 1 && j != remap_button &&
                remap_button != RARCH_UNMAPPED)
            {
               BIT256_SET(handle->buttons, remap_button);
               RARCH_LOG("remap i:%d j:%d n:%d\n", i, j, remap_button);
            }
         }
#if 0
         /* --CURRENTLY NOT IMPLEMENTED--
            this loop should iterate on all users and all analog stick axes and if the axes are
            moved and is assigned to a button it should set the bit on the mapper input bitmap.
            Once implemented we should make sure to clear the original analog 
            stick input in input_state in input_driver.c */

         for (j = RARCH_FIRST_CUSTOM_BIND; j < RARCH_CUSTOM_BIND_LIST_END; j++)
         {

         }
#endif
      }
   }
```

So that portion has to be implemented.

**Comment by Bibalox** [comment]
@fr500  I’m out of the project since a pretty long time. You seem to have a lot of good ideas, i’m pretty sure you will handle this perfectly :)

**Comment by twinaphex** [comment]
@Bibalox He is not a graphical designer though.

**Comment by i30817** [comment]
I'd add that sometimes you want a negation operator for button mapping. For example for the Resident Evil series in dolphin, it's easy and functional to make the characters 'always run' by placing:
x: !key on the mapping for the game, so that characters only walk if you press the button instead of always walk unless you press the button to run.

This has the risks of obvious side effects on other parts of the game interface but for some games it works well enough.

**Comment by twinaphex** [comment]
The lack of button combos is definitely a big omission and it's definitely something I'd like to see in RetroArch, but we just aren't there yet, and I think it also falls outside the scope of this particular PR.

The input system can still use plenty of quality of life improvements in RetroArch, I am not saying things are perfect there, far from it. The input system and the menu system in RetroArch I think are two areas that are showing their age and definitely need heavy work performed on them.

**Comment by i30817** [comment]
I hear you, i also think this PR is already a big project. Just something to keep in mind, since RA approach allows conveniences to be spread to (nearly) all cores.

**Comment by fr500** [comment]
Actually that's perfectly doable (negation) With the groundwork laid out
here.

Basically changes the whole remapping concept to a flexible rules based
remapping system. If input can be read then rules can be written to produce
a desired output.

**Comment by i30817** [comment]
I edited my original post to clarify that a combo system would be what i'd desire, not just negation, because that would be both more generally useful and also have nearly the same requirements. 

However, i do think 'negation' is special and might trip you up if you decide to support it on a combo system.

Simply because it's testing the *absence* of a key being pressed... so the 'real key' _not_ being pressed activates the 'retropad key'... so such a combo would have to be checked every frame if the 'negated' retropad keys in the combo start the combo.

Annoying.

**Comment by rsn8887** [comment]
There's a handle->port, I wonder if you could use it? If handle->port is zero, set buttons for first player, if handle->port is one, set buttons for second player etc.

You are right, input_keys_pressed has hardcoded port to zero everywhere, it only reads player 1 (Why?).

A different function is needed to read the input before all the re-mapping is done. The existing code is just too convoluted and complicated for me to understand at the moment.

**Comment by fr500** [comment]
Yep, but when I dehardcoded the port..... BOOM segfaults everywhere :P

My code is convoluted? the input_mapper.c code seems quite straightforward, but I'm open to criticism, The menu code is what it is, it's just following the established conventions

I think I'm gonna have write an auxiliary function to get the state of each player instead of leveraging that input_keys_pressed one, once I get the state of each it should be trivia to fake the input.

**Comment by fr500** [comment]
@i30817 regarding combos... it's not sad, someone has to take the time to do it, all these things take time and dedication from someone. Everyone has different priorities.

The work here **has to be done** even if this PR flops someone will have to rewrite the input_poll / input_state into something more flexible if you want to have any kind of fakeinput.

FWIW I'm not implementing combos, at least not right now. what I've been doing here is the required changes to the input subsystem so someone can accommodate such a feature later.

**Comment by twinaphex** [comment]
@rsn8887 I agree that it could be a lot more understandable unfortunately. Hopefully we can get there through some refactoring and reworking of some of the code eventually.

**Comment by kivutar** [comment]
I've been talking with radius a little on IRC, but I add my opinion here for the record.

I think we shouldn't display two pads images at the same time. It is confusing. All the other platforms that I can remember (OpenEMU, Steam) chose to display only the *target* pad image for a reason:

 - creating one image per joypad autoconfig is definitively too much work
 - some joypad autoconfig applies for joypads that have different shapes, it can happen for some clones
 - it wouldn't work for joypads that we don't auto detect (manually binded joypads) or would require even more work

I would go even further and say that maintaining 1 joypad image and button layers per emulator is also too much work. In fact, I would display the retropad only in this new binding screen.

I know @fr500 that you really see some advantages in displaying the two exact pads at the same time. It's obviously useful in the case of the genesis pad where buttons are not aligned on a cross. But if we go that way, we will end up with a UI that is always in a half finished state. Like our translations, or like our alternative icon themes.

I would favor a simpler implementation that lowers the maintenance cost over this:

screen shot 2018-04-04 at 9 43 45 am

**Comment by fr500** [comment]
Well that's input binding though, you're only thinking about ONE PAD there.
That looks amazing though :)

**Comment by kivutar** [comment]
OK I forgot about the remappings once again. I will try to come with a mockup with two pads that would apply for remapping. Still, we need to find a solution for unknown pads.

**Comment by kivutar** [comment]
Proposal with a modal during remap:

screen shot 2018-04-04 at 6 18 49 pm

The user have to:
 - press OK on the entry to display the modal
 - use left and right to select the next possible value for this setting

**Comment by fr500** [comment]
Finally fixed P2+ remapping by using libretro_input_binds directly

https://github.com/libretro/RetroArch/pull/6485/commits/c1b49d7e794532b9f4ff03cf936ad349841a4dfb#diff-7797ae0e7813aded66f109c460a562daL1147

Such a stoopid thing :P

**Pull request merged**